front/pages/api/w/[wId]/secrets/index.ts

```diff
@@ -0,0 +1,134 @@
+import type { DustAppSecretType, WithAPIErrorReponse } from "@dust-tt/types";
+import { decrypt, encrypt } from "@dust-tt/types";
+import type { NextApiRequest, NextApiResponse } from "next";
+
+import { Authenticator, getSession } from "@app/lib/auth";
+import { DustAppSecret } from "@app/lib/models/workspace";
+import { apiError, withLogging } from "@app/logger/withlogging";
+
+export type GetSecretsResponseBody = {
+  secrets: DustAppSecretType[];
+};
+
+export type PostSecretsResponseBody = {
+  secret: DustAppSecretType;
+};
+
+async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse<
+    WithAPIErrorReponse<GetSecretsResponseBody | PostSecretsResponseBody>
+  >
+): Promise<void> {
+  const session = await getSession(req, res);
+  const auth = await Authenticator.fromSession(
+    session,
+    req.query.wId as string
+  );
+
+  const owner = auth.workspace();
+  const user = auth.user();
+  if (!owner || !user) {
+    return apiError(req, res, {
+      status_code: 400,
+      api_error: {
+        type: "workspace_not_found",
+        message: "The workspace or user is missing.",
+      },
+    });
+  }
+
+  if (!auth.isBuilder()) {
+    return apiError(req, res, {
+      status_code: 403,
+      api_error: {
+        type: "app_auth_error",
+        message: "You do not have the required permissions.",
+      },
+    });
+  }
+
+  switch (req.method) {
+    case "GET":
+      const secrets = await DustAppSecret.findAll({
+        where: {
+          workspaceId: owner.id,
+          status: "active",
+        },
+        order: [["name", "DESC"]],
+      });
+
+      res.status(200).json({
+        secrets: secrets.map((s) => {
+          const clearSecret = decrypt(s.hash, owner.sId);
+          return {
+            createdAt: s.createdAt.getTime(),
+            name: s.name,
+            value: clearSecret,
+          };
+        }),
+      });
+      return;
+
+    case "DELETE":
+      const { name: deleteSecretName } = req.body;
+      const secret = await DustAppSecret.findOne({
+        where: {
+          name: deleteSecretName,
+          workspaceId: owner.id,
+          status: "active",
+        },
+      });
+
+      if (!secret) {
```

Comment:
We should silently return 204 here as otherwise it can be used to brute-force a secret (which is doable if simple). We should also add rate limits on these endpoints :+1:

Reply:
Done, added rateLimiter and moved to 204